drivers/stepper/A4988.zig

```diff
@@ -0,0 +1,395 @@
+};
+
+// TODO: Make generic? enum for driver e.g. A4988, DRV8824
+pub const A4988 = struct {
+    const Self = @This();
+    microsteps: u8 = 1,
+    ms1_pin: ?mdf.base.Digital_IO,
+    ms2_pin: ?mdf.base.Digital_IO,
+    ms3_pin: ?mdf.base.Digital_IO,
+    dir_pin: mdf.base.Digital_IO,
+    step_pin: mdf.base.Digital_IO,
+    enable_pin: ?mdf.base.Digital_IO,
+    enable_active_state: mdf.base.Digital_IO.State = .low,
+    clock: mdf.base.Clock_Device,
+    rpm: f64 = 0,
+
+    // Movement state
+    profile: Speed_Profile = .constant_speed,
+    accel: u16 = 1000,
+    decel: u16 = 1000,
+
+    steps_to_cruise: u32 = 0,
+    steps_remaining: u32 = 0,
+    steps_to_brake: u32 = 0,
+    step_pulse: u32 = 0,
+    cruise_step_pulse: u32 = 0,
+    rest: u32 = 0,
+    last_action_end: u32 = 0,
+    next_action_interval: u32 = 0,
+    step_count: u32 = 0,
+    // TODO: Check this
+    dir_state: mdf.base.Digital_IO.State = .low,
+
+    const MAX_MICROSTEP = 16;
+    const MOTOR_STEPS = 200;
+    // tA STEP minimium, HIGH pulse width (1us)
+    const STEP_HIGH_MIN = 1;
+    // tB STEP minimium, LOW pulse width (1us)
+    const STEP_LOW_MIN = 1;
+    // t_c = 200ns, setup to change STEP
+    // t_d = 200ns, hold time to change STEP
+    // wakeup time, nSLEEP inactive to STEP (1000us)
+    const wakeup_time = 1000;
+
+    // TODO: Error type?
+    pub fn init(opts: A4988_Options) Self {
+        // TODO: Validate args, optional ms pins
+        // TODO: If any ms pin is set, all must be set
+        return Self{
+            .clock = opts.clock_device,
+            .ms1_pin = opts.ms1_pin,
+            .ms2_pin = opts.ms2_pin,
+            .ms3_pin = opts.ms3_pin,
+            .dir_pin = opts.dir_pin,
+            .step_pin = opts.step_pin,
+            .enable_pin = opts.enable_pin,
+        };
+    }
+
+    pub fn begin(self: *Self, rpm: f64, microstep: u8) !void {
+        // TODO: If ALL ms pins are connected, set to output
+        try self.dir_pin.set_direction(.output);
+        try self.dir_pin.write(.high);
+        try self.step_pin.set_direction(.output);
+        try self.step_pin.write(.low);
+
+        if (self.enable_pin) |pin| {
+            try pin.set_direction(.output);
+            try self.disable();
+        }
+
+        self.rpm = rpm;
+        _ = self.set_microstep(microstep) catch {};
+
+        try self.enable();
+    }
+
+    pub fn enable(self: *Self) !void {
+        if (self.enable_pin) |pin| {
+            try pin.write(self.enable_active_state);
+            self.clock.sleep_us(2);
+        }
+    }
+
+    pub fn disable(self: *Self) !void {
+        if (self.enable_pin) |pin| {
+            try pin.write(if (self.enable_active_state == .high) .low else .high);
+        }
+    }
+
+    pub fn set_rpm(self: Self, rpm: f64) void {
+        self.rpm = rpm;
+    }
+
+    pub fn set_microstep(self: *Self, microsteps: u8) !u8 {
+        // If any MS pins are not defined, return an error and don't change anything
```

Comment:
hm: I think we want to set the struct attribute either way, for when the MS pins are strapped

Reply:
done, but a bit janky, by having `init` call `init_microstep` to 'force' the attribute to the correct value, but then calling `set_microstep` which would return an ignore error.

Basically, if the pins are not defined, we want microsteps set only in `init`, so we know how many steps to expect, but not allow them to be changed (since they are strapped).